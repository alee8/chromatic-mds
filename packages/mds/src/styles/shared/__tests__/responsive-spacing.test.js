/**
 * @jest-environment jsdom
 */
import PropTypes from 'prop-types';
import React from 'react';
import {
  act,
  render,
} from '@testing-library/react';

import ResponsiveBreakpointsProvider, {
  useMdsDeviceSize,
} from '~mds/providers/ResponsiveBreakpoints/ResponsiveBreakpoints';

import {
  BOTTOM,
  CUSTOM_SPACING,
  TOP,
} from '~mds/styles/shared/custom-spacing';
import {
  createMockMatchMedia,
  MOCK_INIT_MQL,
  updateMatchMedia,
} from '~tests/setupTests/media-query-list-helpers';
import {
  SPACING_SIZE_OPTIONS,
  SPACING_SIZE_VALUES,
} from '~mds/constants/prop-types';

import { HeadingSpacing } from './Examples';

import {
  getMarginSpacing,
  getPaddingSpacing,
} from '../responsive-spacing-styles';

const MARGIN_STRING = 'Margin';
const PADDING_STRING = 'Padding';

let originalMatchMedia;
beforeAll(() => {
  originalMatchMedia = window.matchMedia;
  // Mock window functions used in the Header to avoid errors.
  window.matchMedia = createMockMatchMedia(MOCK_INIT_MQL);
  window.scrollTo = jest.fn();
});
afterAll(() => window.matchMedia = originalMatchMedia);

function TestComponentMarginBottom({ device, marginBottom }) {
  const {
    marginBottomClass,
  } = getMarginSpacing({
    marginBottom,
  });
  const isDevice = useMdsDeviceSize();
  const content = isDevice[`is${device}`] ? CUSTOM_SPACING[device.toLowerCase()][marginBottom] : 'cannotDetermine';
  return (
    <ResponsiveBreakpointsProvider>
      <div
        className={marginBottomClass}
        data-testid="marginBottom"
      >
        {content}
      </div>
    </ResponsiveBreakpointsProvider>
  );
}
TestComponentMarginBottom.propTypes = {
  device: PropTypes.string.isRequired,
  marginBottom: PropTypes.string.isRequired,
};

function TestComponentMarginTop({ device, marginTop }) {
  const {
    marginTopClass,
  } = getMarginSpacing({
    marginTop,
  });
  const isDevice = useMdsDeviceSize();
  const content = isDevice[`is${device}`] ? CUSTOM_SPACING[device.toLowerCase()][marginTop] : 'cannotDetermine';
  return (
    <ResponsiveBreakpointsProvider>
      <div
        className={marginTopClass}
        data-testid="marginTop"
      >
        {content}
      </div>
    </ResponsiveBreakpointsProvider>
  );
}
TestComponentMarginTop.propTypes = {
  device: PropTypes.string.isRequired,
  marginTop: PropTypes.string.isRequired,
};

function TestComponentPaddingBottom({ device, paddingBottom }) {
  const {
    paddingBottomClass,
  } = getPaddingSpacing({
    paddingBottom,
  });
  const isDevice = useMdsDeviceSize();
  const content = isDevice[`is${device}`] ? CUSTOM_SPACING[device.toLowerCase()][paddingBottom] : 'cannotDetermine';
  return (
    <ResponsiveBreakpointsProvider>
      <div
        className={paddingBottomClass}
        data-testid="paddingBottom"
      >
        {content}
      </div>
    </ResponsiveBreakpointsProvider>
  );
}
TestComponentPaddingBottom.propTypes = {
  device: PropTypes.string.isRequired,
  paddingBottom: PropTypes.string.isRequired,
};

function TestComponentPaddingTop({ device, paddingTop }) {
  const {
    paddingTopClass,
  } = getPaddingSpacing({
    paddingTop,
  });
  const isDevice = useMdsDeviceSize();
  const content = isDevice[`is${device}`] ? CUSTOM_SPACING[device.toLowerCase()][paddingTop] : 'cannotDetermine';
  return (
    <ResponsiveBreakpointsProvider>
      <div
        className={paddingTopClass}
        data-testid="paddingTop"
      >
        {content}
      </div>
    </ResponsiveBreakpointsProvider>
  );
}
TestComponentPaddingTop.propTypes = {
  device: PropTypes.string.isRequired,
  paddingTop: PropTypes.string.isRequired,
};

// Tests
const SimpleTests = [
  {
    device: 'Desktop', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.desktop, width: 992,
  },
  {
    device: 'Desktop', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.desktop, width: 1000,
  },
  {
    device: 'Desktop', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.desktop, width: 1250,
  },
  {
    device: 'Tablet', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.tablet, width: 768,
  },
  {
    device: 'Tablet', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.tablet, width: 900,
  },
  {
    device: 'Mobile', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.mobile, width: 320,
  },
  {
    device: 'Mobile', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.mobile, width: 500,
  },
  {
    device: 'Mobile', sizes: SPACING_SIZE_OPTIONS, spacings: CUSTOM_SPACING.mobile, width: 576,
  },
];

describe('Test getMarginSpacing hook at different browser widths and margin bottom', () => {
  updateMatchMedia(0, MOCK_INIT_MQL); // Set up to a known state
  SimpleTests.forEach((
    {
      device, sizes, spacings, width,
    }
  ) => {
    sizes.forEach(
      (size) => test(`Simple test of ${device} with ${size} at width=${width} should have ${size}${MARGIN_STRING}${BOTTOM}`, () => {
        const marginBottomClass = `${size}${MARGIN_STRING}${BOTTOM}`;
        act(() => updateMatchMedia(width, MOCK_INIT_MQL));
        const { getByTestId } = render(
          <ResponsiveBreakpointsProvider>
            <TestComponentMarginBottom
              device={device}
              marginBottom={size}
            />
          </ResponsiveBreakpointsProvider>
        );
        const element = getByTestId('marginBottom');
        expect(element.getAttribute('class')).toEqual(marginBottomClass);
        expect(element.firstChild.textContent).toEqual(`${spacings[size]}`);
      })
    );
  });
});

describe('Test getMarginSpacing hook at different browser widths and margin top', () => {
  updateMatchMedia(0, MOCK_INIT_MQL); // Set up to a known state
  SimpleTests.forEach((
    {
      device, sizes, spacings, width,
    }
  ) => {
    sizes.forEach(
      (size) => test(`Simple test of ${device} with ${size} at width=${width} should have ${size}${MARGIN_STRING}${TOP}`, () => {
        const marginTopClass = `${size}${MARGIN_STRING}${TOP}`;
        act(() => updateMatchMedia(width, MOCK_INIT_MQL));
        const { getByTestId } = render(
          <ResponsiveBreakpointsProvider>
            <TestComponentMarginTop
              device={device}
              marginTop={size}
            />
          </ResponsiveBreakpointsProvider>
        );
        const element = getByTestId('marginTop');
        expect(element.getAttribute('class')).toEqual(marginTopClass);
        expect(element.firstChild.textContent).toEqual(`${spacings[size]}`);
      })
    );
  });
});

describe('Test getPaddingSpacing hook at different browser widths and padding bottom', () => {
  updateMatchMedia(0, MOCK_INIT_MQL); // Set up to a known state
  SimpleTests.forEach((
    {
      device, sizes, spacings, width,
    }
  ) => {
    sizes.forEach(
      (size) => test(`Simple test of ${device} with ${size} at width=${width} should have ${size}${PADDING_STRING}${BOTTOM}`, () => {
        const paddingBottomClass = `${size}${PADDING_STRING}${BOTTOM}`;
        act(() => updateMatchMedia(width, MOCK_INIT_MQL));
        const { getByTestId } = render(
          <ResponsiveBreakpointsProvider>
            <TestComponentPaddingBottom
              device={device}
              paddingBottom={size}
            />
          </ResponsiveBreakpointsProvider>
        );
        const element = getByTestId('paddingBottom');
        expect(element.getAttribute('class')).toEqual(paddingBottomClass);
        expect(element.firstChild.textContent).toEqual(`${spacings[size]}`);
      })
    );
  });
});

describe('Test getPaddingSpacing hook at different browser widths and padding top', () => {
  updateMatchMedia(0, MOCK_INIT_MQL); // Set up to a known state
  SimpleTests.forEach((
    {
      device, sizes, spacings, width,
    }
  ) => {
    sizes.forEach(
      (size) => test(`Simple test of ${device} with ${size} at width=${width} should have ${size}${PADDING_STRING}${TOP}`, () => {
        const paddingTopClass = `${size}${PADDING_STRING}${TOP}`;
        act(() => updateMatchMedia(width, MOCK_INIT_MQL));
        const { getByTestId } = render(
          <ResponsiveBreakpointsProvider>
            <TestComponentPaddingTop
              device={device}
              paddingTop={size}
            />
          </ResponsiveBreakpointsProvider>
        );
        const element = getByTestId('paddingTop');
        expect(element.getAttribute('class')).toEqual(paddingTopClass);
        expect(element.firstChild.textContent).toEqual(`${spacings[size]}`);
      })
    );
  });
});

describe('Basic HeadingSpacing with getMarginSpacing and getPaddingSpacing Snapshot', () => {
  test('using defaultValues', () => {
    const { container, unmount } = render(
      <ResponsiveBreakpointsProvider>
        <HeadingSpacing
          marginBottom={SPACING_SIZE_VALUES.xs}
          marginTop={SPACING_SIZE_VALUES.md}
          paddingBottom={SPACING_SIZE_VALUES.xs}
          paddingTop={SPACING_SIZE_VALUES.md}
        />
      </ResponsiveBreakpointsProvider>
    );

    expect(container).toMatchSnapshot();
    unmount(container);
  });
});
