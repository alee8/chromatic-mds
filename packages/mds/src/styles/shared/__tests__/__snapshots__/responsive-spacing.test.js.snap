// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic HeadingSpacing with getMarginSpacing and getPaddingSpacing Snapshot using defaultValues 1`] = `
<div>
  <hr />
  <div
    class="gf-gfBoxWrapper-0-2-1 gf-wrenchBox-0-2-7 gf-wrenchBox-d0-0-2-8 gf-wrenchSpacing-0-2-5 gf-wrenchSpacing-d0-0-2-6"
  >
    <div
      class="gf-mdsGridPlusContainer-0-2-16 gf-mdsGridPlusContainer-d5-0-2-18"
      id=""
    >
      <div
        class="gf-mdsGridPlus-0-2-17 gf-mdsGridPlus-d11-0-2-19 gf-addFlex-0-2-14 gf-addFlex-d0-0-2-15"
      >
        <div
          class="gf-mdsGridPlusItem-0-2-22 gf-mdsGridPlusItem-d0-0-2-23 gf-addFlex-0-2-20 gf-addFlex-d0-0-2-21"
        >
          <div
            class="xsMarginBottom mdMarginTop xsPaddingBottom mdPaddingTop gf-heading-0-2-9 gf-heading-d0-0-2-12  gf-headingSize-0-2-10 gf-headingSize-d1-0-2-13  "
            id=""
          >
            <h1>
              Responsive spacing for gf
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    class="mntv-mntvBoxWrapper-0-2-2 mntv-wrenchBox-0-2-7 mntv-wrenchBox-d0-0-2-8 mntv-wrenchSpacing-0-2-5 mntv-wrenchSpacing-d0-0-2-6"
  >
    <div
      class="mntv-mdsGridPlusContainer-0-2-16 mntv-mdsGridPlusContainer-d5-0-2-18"
      id=""
    >
      <div
        class="mntv-mdsGridPlus-0-2-17 mntv-mdsGridPlus-d11-0-2-19 mntv-addFlex-0-2-14 mntv-addFlex-d0-0-2-15"
      >
        <div
          class="mntv-mdsGridPlusItem-0-2-22 mntv-mdsGridPlusItem-d0-0-2-23 mntv-addFlex-0-2-20 mntv-addFlex-d0-0-2-21"
        >
          <div
            class="xsMarginBottom mdMarginTop xsPaddingBottom mdPaddingTop mntv-heading-0-2-9 mntv-heading-d0-0-2-12  mntv-headingSize-0-2-10 mntv-headingSize-d1-0-2-13  "
            id=""
          >
            <h1>
              Responsive spacing for mntv
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    class="smHelp-smHelpBoxWrapper-0-2-3 smHelp-wrenchBox-0-2-7 smHelp-wrenchBox-d0-0-2-8 smHelp-wrenchSpacing-0-2-5 smHelp-wrenchSpacing-d0-0-2-6"
  >
    <div
      class="smHelp-mdsGridPlusContainer-0-2-16 smHelp-mdsGridPlusContainer-d5-0-2-18"
      id=""
    >
      <div
        class="smHelp-mdsGridPlus-0-2-17 smHelp-mdsGridPlus-d11-0-2-19 smHelp-addFlex-0-2-14 smHelp-addFlex-d0-0-2-15"
      >
        <div
          class="smHelp-mdsGridPlusItem-0-2-22 smHelp-mdsGridPlusItem-d0-0-2-23 smHelp-addFlex-0-2-20 smHelp-addFlex-d0-0-2-21"
        >
          <div
            class="xsMarginBottom mdMarginTop xsPaddingBottom mdPaddingTop smHelp-heading-0-2-9 smHelp-heading-d0-0-2-12 smHelp-headingWithLink-0-2-11 smHelp-headingSize-0-2-10 smHelp-headingSize-d1-0-2-13  "
            id=""
          >
            <h1>
              Responsive spacing for smHelp
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    class="sm-smBoxWrapper-0-2-4 sm-wrenchBox-0-2-7 sm-wrenchBox-d0-0-2-8 sm-wrenchSpacing-0-2-5 sm-wrenchSpacing-d0-0-2-6"
  >
    <div
      class="sm-mdsGridPlusContainer-0-2-16 sm-mdsGridPlusContainer-d5-0-2-18"
      id=""
    >
      <div
        class="sm-mdsGridPlus-0-2-17 sm-mdsGridPlus-d11-0-2-19 sm-addFlex-0-2-14 sm-addFlex-d0-0-2-15"
      >
        <div
          class="sm-mdsGridPlusItem-0-2-22 sm-mdsGridPlusItem-d0-0-2-23 sm-addFlex-0-2-20 sm-addFlex-d0-0-2-21"
        >
          <div
            class="xsMarginBottom mdMarginTop xsPaddingBottom mdPaddingTop sm-heading-0-2-9 sm-heading-d0-0-2-12 sm-headingWithLink-0-2-11 sm-headingSize-0-2-10 sm-headingSize-d1-0-2-13  "
            id=""
          >
            <h1>
              Responsive spacing for sm
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
</div>
`;
